### ISSUE TITLE
Rule suggestion: forbid use of str[idx] syntax (index string as array of UTF-16 code units)

### ISSUE THREAD

**Issue by NWilson** [typescript-eslint/typescript-eslint#2927]
**Repro**

I would like a rule to prevent use of the array-index syntax on string-typed values.
- These slice UTF-16 code sequences, so codePointAt is preferable
- The use of array syntax makes it impossible to search for these using regex (there is no way to narrow a manual regex search to detect use of index syntax only on string values); however a type-aware linter can detect these

```TS
function test(str: string) {
  return str[0]; // !! should disallow
}
```

**Comment by JoshuaKGoldberg** [comment]
Err @NWilson @bradzacher now I'm not so sure. `charPointAt` and `codePointAt` give numbers ([SO explanation of charPointAt and codePointAt](https://stackoverflow.com/questions/36527642/difference-between-codepointat-and-charcodeat)) and `[0]` is functionally similar to `.at(0)` ([SO comparison of at vs index](https://stackoverflow.com/questions/5943726/string-charatx-or-stringx)).

This seems like a rather niche rule. Is there much benefit to including it in the core typescript-eslint ruleset? Is there a code snippet that can demonstrate better behavior using, say, `at()` vs `[]`?

**Comment by bradzacher** [comment]
I think this is pretty niche. It's rare that anyone ends to do single-character access on a string, and it's even rarer that you're doing that access in conjunction with strings that much contain multi-byte characters.

I don't see that there is enough broad applicability for this to exist within our general-purpose plugin.

Additionally there's nothing specific to typescript in this request - it's just a rule that requires type-information.

For the above reasons, I think we should reject this request for our project.

Doesn't mean it's not a good idea! Just that I think it should be built and maintained in an external plugin.

**Comment by JoshuaKGoldberg** [comment]
Agreed. If the plugin ends up becoming popular we can always re-evaluate adding it into this project. Thanks!
